### PR TITLE
Transition to TGS Soft impulse solver

### PR DESCRIPTION
**Pull request by fallenatlas**
# Description

Oh boy...

Transition from XPBD solver to TGS solver.

Much of the solver for the penetration constraints was changed, we now make corrections based on impulses, based mainly on the equations for conservation of momentum with some bias proportional to the penetration which will work to separate the particles.

Collisions also work as soft constraints now, the values might need some tweaking in the future but they seem ok for now.

The integration also had to be changed and is now under the solver, because it just makes more sense.

The new loop is as follows:

- fixed step
  - collision stuff (unchanged)
  - prepare constraints (for now just create the penetration constraint for colliding pairs)
  - fixed substep
    - integrate velocity
    - solve velocity with bias
    - integrate position
    - solve velocity without bias (relax)
  - finalize position (apply position correction)

Currently the impulses are enough only to separate the particles, this means that some collisions may not look as expected because everything is perfectly inelastic. Elastic and inelastic collisions will come with restitution which should be the next PR.

Also of note that stacks of more than 2 particles aren't very good yet, so the complex physics sample won't look the best.

## Checklist

- [x] Self-review changes.
- [x] Evaluate impact on the documentation.
- [x] Add entry to the changelog's unreleased section.

### CODE REVIEW

File: engine/include/cubos/engine/physics/plugin.hpp [hunk 1]
```diff
@@ -39,24 +38,16 @@ namespace cubos::engine
...
     /// - @ref Impulse - holds impulses applied on a particle.
     /// - @ref Mass - holds the mass of an object.
-    /// - @ref PreviousPosition - holds the previous position of the entity in a substep.
     /// - @ref AccumulatedCorrection - holds the corrections accumulated from the constraints solving.
     ///
     /// ## Dependencies
     /// - @ref physics-gravity-plugin
+    ///
+    /// ## Tags
+    /// - physicsApplyForcesTag - tag for all systems that modify velocity or apply forces or impulses.
```
Comment (RiscadoA) [inline engine/include/cubos/engine/physics/plugin.hpp:48]:
```suggestion

    /// @brief Tag which should be used on all systems that modify velocity or apply forces or impulses.
```

File: engine/src/physics/solver/integration/plugin.cpp [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+#include <cubos/engine/transform/plugin.hpp>
+
+#include "../../fixed_substep/plugin.hpp"
+
+using namespace cubos::engine;
+
+CUBOS_DEFINE_TAG(cubos::engine::physicsApplyImpulsesTag);
+CUBOS_DEFINE_TAG(cubos::engine::physicsClearForcesTag);
+
+void cubos::engine::integrationPlugin(Cubos& cubos)
```
Comment (RiscadoA) [inline engine/src/physics/solver/integration/plugin.cpp:18]:
```suggestion
void cubos::engine::physicsIntegrationPlugin(Cubos& cubos)
```

File: engine/samples/complex-physics/main.cpp
Comment (RiscadoA) [inline engine/samples/complex-physics/main.cpp:1]: Just noticed the folder name doesn't follow the convention, it should be complex_physics.

File: engine/src/physics/plugin.cpp [hunk 3]
```diff
@@ -76,9 +61,6 @@ CUBOS_REFLECT_IMPL(Damping)
 
 void cubos::engine::physicsPlugin(Cubos& cubos)
 {
-    cubos.depends(fixedStepPlugin);
-    cubos.depends(transformPlugin);
-
     cubos.plugin(fixedSubstepPlugin);
```
Comment (RiscadoA) [inline engine/src/physics/plugin.cpp:64]: This plugin should be prefixed with `physics`.

**Review by RiscadoA — CHANGES_REQUESTED** [review]
Thanks for working on this! That was very quick :eyes: good job!

**Review by DiogoMendonc-a — APPROVED** [review]
LGTM, but to be fair, I don't really know what I'm looking at for most of these maths 😅

**Review by tomas7770 — APPROVED** [review]
LGTM as far as I can understand!

File: engine/include/cubos/engine/physics/plugin.hpp [hunk 4]
```diff
@@ -39,24 +38,15 @@ namespace cubos::engine
     /// - @ref Force - holds forces applied on a particle.
     /// - @ref Impulse - holds impulses applied on a particle.
     /// - @ref Mass - holds the mass of an object.
-    /// - @ref PreviousPosition - holds the previous position of the entity in a substep.
     /// - @ref AccumulatedCorrection - holds the corrections accumulated from the constraints solving.
     ///
     /// ## Dependencies
     /// - @ref physics-gravity-plugin
+    ///
```
Comment (RiscadoA) [inline engine/include/cubos/engine/physics/plugin.hpp:45]:
```suggestion
```

File: engine/src/physics/solver/plugin.cpp [hunk 5]
```diff
@@ -1,22 +1,58 @@
...
+CUBOS_DEFINE_TAG(cubos::engine::physicsSolveContactTag);
+CUBOS_DEFINE_TAG(cubos::engine::physicsIntegratePositionTag);
+CUBOS_DEFINE_TAG(cubos::engine::physicsSolveRelaxConstraintTag);
+CUBOS_DEFINE_TAG(cubos::engine::physicsSolveRelaxUserConstraintTag);
+CUBOS_DEFINE_TAG(cubos::engine::physicsSolveRelaxContactTag);
+CUBOS_DEFINE_TAG(cubos::engine::physicsFinalizePositionTag);
 
 using namespace cubos::engine;
 
 void cubos::engine::solverPlugin(Cubos& cubos)
```
Comment (RiscadoA) [inline engine/src/physics/solver/plugin.cpp:24]:
```suggestion
void cubos::engine::physicsSolverPlugin(Cubos& cubos)
```

**Review by RiscadoA — APPROVED** [review]
Other than the comments I added, LGTM!

**Pull request merged**